crypto/uid.c

```diff
@@ -44,10 +49,50 @@ int OPENSSL_issetugid(void)
 #  endif
 # endif
 
+/*
+ * Allows for slightly more permissive environment variable retrieval. Requires capability checks
+ */
+#if defined (OPENSSL_NETCAP_ALLOW_ENV) && defined(__linux__)
+/*
+ * Tests to see if a process has ONLY the requested capability
+ * see kernel/capability.c in the linux kernel source for more details
+ * structs are defined in sys/capability.h
+ */
+int HasOnlyCapability(int capability)
+{
+
+    if (!cap_valid(capability)) {
+        return 0;
+    }
+
+    struct __user_cap_data_struct cap_data[2];
+    struct __user_cap_header_struct cap_header_data = {
+        _LINUX_CAPABILITY_VERSION_3,
+        getpid()};
+
+    if (syscall(SYS_capget, &cap_header_data, &cap_data) != 0) {
+        return 0;
+    }
+
+    if (capability < 32) {
+        return cap_data[0].permitted == (CAP_TO_MASK(capability));
+    }
+    // Probably also need to check [1] - some capabilities are >32
+    // to check for ONLY net_bind we need to ensure these are 0 also
+    
+    return cap_data[1].permitted == (CAP_TO_MASK(capability));
+}
+#endif
+
 int OPENSSL_issetugid(void)
```

Comment:
nit..overloading the bool check this way hides the impl somewhat for OPENSSL_issetugid when both OSSL_IMPLEMENT_GETAUXVAL and OPENSSL_NETCAP_ALLOW_ENV are defined.. suggest keeping use of HasOnlyCapability(CAP_NET_BIND_SERVICE) separate from the u/gid check or maybe rename OPENSSL_allowlimtedenvcopy(void) ..

Reply:
Thanks, that's a good suggestion.